File: Assignment/CODING/Cau1/Cau1.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.*;

class ERD {
    String name;
    String attributes;
    String primaryKey;

    public ERD(String name, String attributes, String primaryKey) {
        this.name = name;
        this.attributes = attributes;
        this.primaryKey = primaryKey;
    }

    public String toString() {
        return this.name + "(" + this.attributes + ")" + " PK: " + this.primaryKey;
    }
}

class RelationalSchema {
    String name;
    String attributes;
    String primaryKey;
    String foreignKey;

    public RelationalSchema(String name, String attributes, String primaryKey, String foreignKey) {
        this.name = name;
        this.attributes = attributes;
        this.primaryKey = primaryKey;
        this.foreignKey = foreignKey;
    }

    public String toString() {
        String result;
        if (this.foreignKey == "")
            result = this.name + "(" + this.attributes + ")" + " PK: " + this.primaryKey;
        else
            result = this.name + "(" + this.attributes + ")" + " PK: " + this.primaryKey + " FK: " + this.foreignKey;
        return result;
    }
}

class Relationship {
    String firstEntity;
    String secondEntity;
    String type;

    public Relationship(String firstEntity, String secondEntity, String type) {
        this.firstEntity = firstEntity;
        this.secondEntity = secondEntity;
        this.type = type;
    }

    public String toString() {
        return this.firstEntity + "-" + this.secondEntity + ": " + this.type;
    }
}

class ERD2RS {
    public static void ERDToRelationalSchema(String inputPath, String outputPath) {
        List<ERD> database = new ArrayList<>();
        List<Relationship> relationships = new ArrayList<>();
        List<RelationalSchema> resultDatabase = new ArrayList<>();
        List<Relationship> resultRelationships = new ArrayList<>();
        List<String> existed = new ArrayList<>();

        try (BufferedReader myInput = new BufferedReader(new FileReader(inputPath))) {
            String line;
            while ((line = myInput.readLine()) != null) {
                String[] data = line.split(" ");
                if (data.length == 3) {
                    database.add(new ERD(data[0], data[1], data[2]));
                } else if (data.length == 2) {
                    if (!data[0].contains("-")) {
                        // Nếu không có dấu '-' thì đây là thực thể con
                        String entityName = data[0];
                        String attributes = data[1];
                        String primaryKey = "ID"; // Giả sử primaryKey của thực thể con là "ID"
                        database.add(new ERD(entityName, attributes, primaryKey));
                    } else {
                        String[] entities = data[0].split("-");
                        relationships.add(new Relationship(entities[0], entities[1], data[1]));
                    }
                }
            }

            for (Relationship relationship : relationships) {
                for (ERD firstEntity : database) {
                    if (relationship.firstEntity.equals(firstEntity.name)) {

                        // Quan hệ 1-1
                        if (relationship.type.equals("1-1")) {
                            // Khoá chính ở phía bắt buộc làm khoá ngoại ở phía tuỳ chọn
                            String foreignKey = firstEntity.primaryKey;
                            for (ERD secondEntity : database) {
                                if (relationship.secondEntity.equals(secondEntity.name)) {
                                    if (!existed.contains(firstEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(firstEntity.name,
                                                firstEntity.attributes, firstEntity.primaryKey, ""));
                                        existed.add(firstEntity.name);
                                    }
                                    if (!existed.contains(secondEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(secondEntity.name,
                                                secondEntity.attributes + ',' + foreignKey,
                                                secondEntity.primaryKey, foreignKey));
                                        existed.add(secondEntity.name);
                                    }
                                }
                            }
                        }

                        // Quan hệ 1-N
                        if (relationship.type.equals("1-N")) {
                            // Khoá chính ở phía quan hệ một làm khoá ngoại ở phía quan hệ nhiều
                            String foreignKey = firstEntity.primaryKey;
                            for (ERD secondEntity : database) {
                                if (relationship.secondEntity.equals(secondEntity.name)) {
                                    if (!existed.contains(firstEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(firstEntity.name,
                                                firstEntity.attributes, firstEntity.primaryKey, ""));
                                        existed.add(firstEntity.name);
                                    }
                                    if (existed.contains(secondEntity.name)) {
                                        // Trường hợp bảng có nhiều khóa ngoại
                                        for (RelationalSchema entity : resultDatabase) {
                                            if (entity.name.equals(secondEntity.name)) {
                                                entity.attributes += (',' + foreignKey);
                                                entity.foreignKey += (',' + foreignKey);
                                                break;
                                            }
                                        }
                                    }
                                    if (!existed.contains(secondEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(secondEntity.name,
                                                secondEntity.attributes + ',' + foreignKey,
                                                secondEntity.primaryKey, foreignKey));
                                        existed.add(secondEntity.name);
                                    }
                                }
                            }
                        }

                        // Quan hệ N-N
                        else if (relationship.type.equals("N-N")) {
                            for (ERD secondEntity : database) {
                                if (relationship.secondEntity.equals(secondEntity.name)) {
                                    System.out.println("This is an N-N relationship between entities  " +
                                            firstEntity.name + " and " + secondEntity.name);
                                    String nameNewTable = System.console().readLine(
                                            "Enter the name for the new relationship table: ");
                                    String attributesNewTable = System.console().readLine(
                                            "Enter attributes for the new table: ");
                                    resultDatabase.add(new RelationalSchema(nameNewTable,
                                            firstEntity.primaryKey + ',' + secondEntity.primaryKey + ','
                                                    + attributesNewTable,
                                            firstEntity.primaryKey + ',' + secondEntity.primaryKey,
                                            firstEntity.primaryKey + ',' + secondEntity.primaryKey));
                                    existed.add(nameNewTable);
                                    resultRelationships.add(new Relationship(
                                            firstEntity.name, nameNewTable, "1-N"));
                                    resultRelationships.add(new Relationship(
                                            secondEntity.name, nameNewTable, "1-N"));

                                    if (!existed.contains(firstEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(firstEntity.name,
                                                firstEntity.attributes, firstEntity.primaryKey, ""));
                                        existed.add(firstEntity.name);
                                    }

                                    if (!existed.contains(secondEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(secondEntity.name,
                                                secondEntity.attributes, secondEntity.primaryKey, ""));
                                        existed.add(secondEntity.name);
                                    }
                                }
                            }
                        }

                        // Quan hệ kế thừa
                        else if (relationship.type.equals("Inheritance")) {
                            // Khoá chính của quan hệ cha trở thành khoá chính của quan hệ con
                            String primaryKey = firstEntity.primaryKey;
                            for (ERD secondEntity : database) {
                                if (relationship.secondEntity.equals(secondEntity.name)) {
                                    if (!existed.contains(firstEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(firstEntity.name,
                                                firstEntity.attributes, firstEntity.primaryKey, ""));
                                        existed.add(firstEntity.name);
                                    }
                                    if (!existed.contains(secondEntity.name)) {
                                        resultDatabase.add(new RelationalSchema(secondEntity.name,
                                                secondEntity.attributes + ',' + primaryKey,
                                                primaryKey, primaryKey));
                                        existed.add(secondEntity.name);
                                    }
                                }
                            }
                        }

                        // Quan hệ thực thể mạnh - thực thể yếu
                        else if (relationship.type.equals("Strong-Weak")) {
                            String foreignKey = firstEntity.primaryKey;
                            for (ERD secondEntity : database) {
                                if (!existed.contains(firstEntity.name)) {
                                    resultDatabase.add(new RelationalSchema(firstEntity.name,
                                            firstEntity.attributes, firstEntity.primaryKey, ""));
                                    existed.add(firstEntity.name);
                                }
                                if (!existed.contains(secondEntity.name)) {
                                    resultDatabase.add(new RelationalSchema(secondEntity.name,
                                            secondEntity.attributes + ',' + foreignKey,
                                            secondEntity.primaryKey + ',' +foreignKey,
                                            foreignKey));
                                }
                            }
                        }
                    }
                }
            }

            try (BufferedWriter myOutput = new BufferedWriter(new FileWriter(outputPath))) {
                for (RelationalSchema schema : resultDatabase) {
                    myOutput.write(schema.toString());
                    myOutput.newLine();
                }

                for (Relationship relationship : relationships) {
                    myOutput.write(relationship.toString());
                    myOutput.newLine();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        ERDToRelationalSchema("input.txt", "output.txt");
    }
}

File: Assignment/CODING/Cau1/Cau1.py
class ERD:
    def __init__(self, name, attributes, primaryKey):
        self.name = name
        self.attributes = attributes
        self.primaryKey = primaryKey

    def __str__(self):
        return f"{self.name}({self.attributes}) PK: {self.primaryKey}"


class RelationalSchema:
    def __init__(self, name, attributes, primaryKey, foreignKey):
        self.name = name
        self.attributes = attributes
        self.primaryKey = primaryKey
        self.foreignKey = foreignKey

    def __str__(self):
        if self.foreignKey == "":
            return f"{self.name}({self.attributes}) PK:{self.primaryKey}"
        else:
            return f"{self.name}({self.attributes}) PK:{self.primaryKey} FK:{self.foreignKey}"


class Relationship:
    def __init__(self, firstEntity, secondEntity, type):
        self.firstEntity = firstEntity
        self.secondEntity = secondEntity
        self.type = type

    def __str__(self):
        return f"{self.firstEntity}-{self.secondEntity}: {self.type}"


def ERD2RS(inputPath, outputPath):
    Entities = [] # Khởi tạo mảng chứ các thực thể
    Relationships = [] # Khởi tạo mảng chứ các quan hệ giữa các thực thể
    Table = [] # Khởi tạo mảng chứ các table đã được chuyển từ ERD sang RelationalSchema
    AERelationship = [] # Khởi tạo mảng chứ các quan hệ với thực thể kết hợp
    TableExisted = [] # Khởi tạo mảng chứ tên các table đã được khởi tạo

    with open(inputPath, 'r') as myInput:
        for line in myInput:
            data = line.split()
            if len(data) == 3:
                Entities.append(ERD(data[0], data[1], data[2]))
            elif len(data) == 2:
                if '-' not in data[0]:
                    entityName = data[0]
                    attributes = data[1]
                    primaryKey = ""
                    Entities.append(ERD(entityName, attributes, primaryKey))
                else:
                    entities = data[0].split("-")
                    Relationships.append(Relationship(entities[0], entities[1], data[1]))

        for relationship in Relationships:
            for firstEntity in Entities:
                if relationship.firstEntity == firstEntity.name:

                    # Quan hệ 1-1
                    if relationship.type == "1-1":
                        foreignKey = firstEntity.primaryKey
                        for secondEntity in Entities:
                            if relationship.secondEntity == secondEntity.name:
                                if firstEntity.name not in TableExisted:
                                    Table.append(RelationalSchema(firstEntity.name,
                                                                           firstEntity.attributes,
                                                                           firstEntity.primaryKey, ""))
                                    TableExisted.append(firstEntity.name)
                                if secondEntity.name not in TableExisted:
                                    Table.append(RelationalSchema(secondEntity.name,
                                                                           secondEntity.attributes + ',' + foreignKey,
                                                                           secondEntity.primaryKey, foreignKey))
                                    TableExisted.append(secondEntity.name)

                    # Quan hệ 1-N
                    if relationship.type == "1-N":
                        foreignKey = firstEntity.primaryKey
                        for secondEntity in Entities:
                            if relationship.secondEntity == secondEntity.name:
                                if firstEntity.name not in TableExisted:
                                    Table.append(RelationalSchema(firstEntity.name,
                                                                           firstEntity.attributes,
                                                                           firstEntity.primaryKey, ""))
                                    TableExisted.append(firstEntity.name)
                                if secondEntity.name in TableExisted:
                                    for entity in Table:
                                        if entity.name == secondEntity.name:
                                            entity.attributes += ',' + foreignKey
                                            entity.foreignKey += ',' + foreignKey
                                            break
                                else:
                                    Table.append(RelationalSchema(secondEntity.name,
                                                                           secondEntity.attributes + ',' + foreignKey,
                                                                           secondEntity.primaryKey, foreignKey))
                                    TableExisted.append(secondEntity.name)

                    # Quan hệ N-N
                    if relationship.type == 'N-N':
                        for secondEntity in Entities:
                            if relationship.secondEntity == secondEntity.name:
                                print("Đây là mối quan hệ N-N giữa thực thể " +
                                    firstEntity.name + " và thực thể " + secondEntity.name)
                                nameNewTable = input(
                                    "Vui lòng nhập tên quan hệ mới được tạo ra từ quan hệ N-N trên: ")
                                attributesNewTable = input(
                                    "Vui lòng nhập các thuộc tính muốn thêm cho quan hệ này: ")
                                Table.append(RelationalSchema(nameNewTable, firstEntity.primaryKey + ',' + secondEntity.primaryKey + ',' + attributesNewTable,
                                                                    firstEntity.primaryKey + ',' + secondEntity.primaryKey, firstEntity.primaryKey + ',' + secondEntity.primaryKey))
                                TableExisted.append(nameNewTable)
                                AERelationship.append(Relationship(
                                    firstEntity.name, nameNewTable, "1-N"))
                                AERelationship.append(Relationship(
                                    secondEntity.name, nameNewTable, "1-N"))

                    # Quan hệ kế thừa
                    if relationship.type == "Inheritance":
                        primaryKey = firstEntity.primaryKey
                        for secondEntity in Entities:
                            if relationship.secondEntity == secondEntity.name:
                                if firstEntity.name not in TableExisted:
                                    Table.append(RelationalSchema(firstEntity.name,
                                                                           firstEntity.attributes,
                                                                           firstEntity.primaryKey, ""))
                                    TableExisted.append(firstEntity.name)
                                if secondEntity.name not in TableExisted:
                                    Table.append(RelationalSchema(secondEntity.name,
                                                                           secondEntity.attributes + ',' + primaryKey,
                                                                           primaryKey, primaryKey))
                                    TableExisted.append(secondEntity.name)

                    # Quan hệ thực thể mạnh-yếu
                    if relationship.type == "Strong-Weak":
                        foreignKey = firstEntity.primaryKey
                        for secondEntity in Entities:
                            if secondEntity.name not in TableExisted:
                                Table.append(RelationalSchema(secondEntity.name,
                                                                       secondEntity.attributes + ',' + foreignKey,
                                                                       secondEntity.primaryKey + ',' + foreignKey,
                                                                       foreignKey))
                                TableExisted.append(secondEntity.name)

    with open(outputPath, 'w', encoding='utf-8') as myOutput:
        myOutput.write("Các bảng:\n")
        for schema in Table:
            myOutput.write(str(schema) + '\n')

        myOutput.write("\nCác mối quan hệ:\n")
        for relationship in Relationships:
            myOutput.write(str(relationship) + '\n')



ERD2RS("input.txt", "output.txt")
